cluster/reconcile.go

```diff
@@ -243,3 +250,69 @@ func cleanControlNode(ctx context.Context, kubeCluster, currentCluster *Cluster,
 	}
 	return nil
 }
+
+func restartComponentsWhenCertChanges(ctx context.Context, currentCluster, kubeCluster *Cluster) error {
+	AllCertsMap := map[string]bool{
+		pki.KubeAPICertName:            false,
+		pki.RequestHeaderCACertName:    false,
+		pki.CACertName:                 false,
+		pki.ServiceAccountTokenKeyName: false,
+		pki.APIProxyClientCertName:     false,
+		pki.KubeControllerCertName:     false,
+		pki.KubeSchedulerCertName:      false,
+		pki.KubeProxyCertName:          false,
+		pki.KubeNodeCertName:           false,
+	}
+	checkCertificateChanges(ctx, currentCluster, kubeCluster, AllCertsMap)
+	// check Restart Function
+	allHosts := hosts.GetUniqueHostList(kubeCluster.EtcdHosts, kubeCluster.ControlPlaneHosts, kubeCluster.WorkerHosts)
+	AllCertsFuncMap := map[string][]services.RestartFunc{
+		pki.CACertName:                 []services.RestartFunc{services.RestartKubeAPI, services.RestartKubeController, services.RestartKubelet},
+		pki.KubeAPICertName:            []services.RestartFunc{services.RestartKubeAPI, services.RestartKubeController},
+		pki.RequestHeaderCACertName:    []services.RestartFunc{services.RestartKubeAPI},
+		pki.ServiceAccountTokenKeyName: []services.RestartFunc{services.RestartKubeAPI, services.RestartKubeController},
+		pki.APIProxyClientCertName:     []services.RestartFunc{services.RestartKubeAPI},
+		pki.KubeControllerCertName:     []services.RestartFunc{services.RestartKubeController},
+		pki.KubeSchedulerCertName:      []services.RestartFunc{services.RestartScheduler},
+		pki.KubeProxyCertName:          []services.RestartFunc{services.RestartKubeproxy},
+		pki.KubeNodeCertName:           []services.RestartFunc{services.RestartKubelet},
+	}
+	for certName, changed := range AllCertsMap {
+		if changed {
+			for _, host := range allHosts {
+				for _, restartFunc := range AllCertsFuncMap[certName] {
```

Comment:
Can we extract this in it's own function ? maybe we can wrap it in a errgroup for parallel rune.